### PR TITLE
Fix out of date documentation & remove friction points

### PR DESCRIPTION
**Pull request by ZainRizvi**
Fixes various friction points with the documentation for onboarding new users and remove instructions that were no longer valid

Changes include:
- Listing prerequisites earlier, so that devs can ensure they're met before encountering error messages
- Removing linter invocations that are no longer valid
- Modifying instructions to install mkl packages to only apply to x86 based CPUs

[skip ci]

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -81,6 +83,13 @@ https://github.com/pytorch/pytorch#from-source
 
 To develop PyTorch on your machine, here are some tips:
 
+### Prerequisites
+* CMake. We recommend installing it with [Homebrew](https://brew.sh/)
+with `brew install cmake` if you are developing on MacOS or Linux system.
```
Comment (seemethere) [inline CONTRIBUTING.md:88]:
```suggestion
with `brew install cmake` if you are developing on MacOS.
```

I'd probably reword this to suggest installing it through their package manager since using brew on a linux box seems like an anti-pattern when compared to macOS

**Review by seemethere — APPROVED** [review]
Looks mostly good just one nit about using homebrew on linux

**Review by janeyx99 — APPROVED** [review]
thanks!

File: CONTRIBUTING.md (hunk 1 above)
Comment (malfet) [inline CONTRIBUTING.md:88]: This seems wrong. CMake should better be installed from [PyPI](https://pypi.org/project/cmake/)

File: README.md [hunk 2]
```diff
@@ -255,7 +268,7 @@ come with Visual Studio Code by default.
 
 If you want to build legacy python code, please refer to [Building on legacy code and CUDA](https://github.com/pytorch/pytorch/blob/master/CONTRIBUTING.md#building-on-legacy-code-and-cuda)
 
-Build with CPU
+**Build with CPU**
```
Comment (malfet) [inline README.md:271]: Can we somehow rename/rephrase it, say "Standard build/generic build, etc" As it is not not possible to build PyTorch without CPU (I would argue one could not even power computer on without CPU"

File: README.md [hunk 3]
```diff
@@ -211,15 +223,18 @@ git submodule update --init --recursive --jobs 0
...
-python setup.py install
+# Only run this if you're compiling for ROCm
+python tools/amd_build/build_amd.py
 ```
 
-Note that if you are compiling for ROCm, you must run this command first:
+Install PyTorch
 ```bash
-python tools/amd_build/build_amd.py
+export CMAKE_PREFIX_PATH=${CONDA_PREFIX:-"$(dirname $(which conda))/../"}
```
Comment (malfet) [inline README.md:236]: Is this really necessary? I never had to pass this option

File: README.md [hunk 4]
```diff
@@ -177,25 +181,33 @@ Other potentially useful environment variables may be found in `setup.py`.
 
 #### Install Dependencies
 
-Common
+**Common**
+
 ```bash
-conda install astunparse numpy ninja pyyaml mkl mkl-include setuptools cmake cffi typing_extensions future six requests dataclasses
```
Comment (malfet) [inline README.md:182]: This sounds too complicated. Can we keep the common installation path, and then add a footnote about AppleM1/aarch64 machines?

File: README.md [hunk 5]
```diff
@@ -152,16 +153,19 @@ They require JetPack 4.2 and above, and [@dusty-nv](https://github.com/dusty-nv)
 
 ### From Source
 
-If you are installing from source, you will need Python 3.7 or later and a C++14 compiler. Also, we highly recommend installing an [Anaconda](https://www.anaconda.com/distribution/#download-section) environment.
-You will get a high-quality BLAS library (MKL) and you get controlled dependency versions regardless of your Linux distro.
+#### Prerequisites
+If you are installing from source, you will need:
+- Python 3.7 or later (for Linux, Python 3.7.6+ or 3.8.1+ is needed)
+- A C++14 compiler, such as cmake
```
Comment (malfet) [inline README.md:159]: cmake is not a compiler, is it?
```suggestion
- A C++14 compatible compiler, such as clang
```

File: README.md [hunk 6]
```diff
@@ -152,16 +153,19 @@ They require JetPack 4.2 and above, and [@dusty-nv](https://github.com/dusty-nv)
 
 ### From Source
 
-If you are installing from source, you will need Python 3.7 or later and a C++14 compiler. Also, we highly recommend installing an [Anaconda](https://www.anaconda.com/distribution/#download-section) environment.
-You will get a high-quality BLAS library (MKL) and you get controlled dependency versions regardless of your Linux distro.
+#### Prerequisites
+If you are installing from source, you will need:
+- Python 3.7 or later (for Linux, Python 3.7.6+ or 3.8.1+ is needed)
```
Comment (malfet) [inline README.md:158]: Err, why? 3.7.0 should work, shouldn't it?

**Review by malfet — CHANGES_REQUESTED** [review]
Thank you for improving the docs, but it seems to need one small iteration

File: CONTRIBUTING.md (hunk 1 above)
Comment (ZainRizvi) [reply CONTRIBUTING.md]: Makes sense. Updated!

File: README.md (hunk 6 above)
Comment (ZainRizvi) [reply README.md:158]: I didn't test this, but from the current Readme.md in master:

image

File: README.md (hunk 5 above)
Comment (ZainRizvi) [reply README.md:159]: uhh...right 😅

File: README.md (hunk 4 above)
Comment (ZainRizvi) [reply README.md:182]: Which part feels complicated? Moving the mlk installation bits to the OS specific sections?

This may make copy/pasting a little more tedious for someone who's regularly using these instructions to install pytorch (they have to copy/paste one extra line now), but for someone who's installing pytorch from source for the first time I would argue that these instructions are simpler.  They just keep going from top to bottom and copy/pasting the commands that apply to them.

File: README.md (hunk 3 above)
Comment (ZainRizvi) [reply README.md:236]: No idea. This was the pre-existing instruction and I didn't want to modify it right now in case there are corner cases where this is needed.

File: README.md (hunk 2 above)
Comment (ZainRizvi) [reply README.md:271]: Yeah, that whole section was a bit weird.  I believe it's talking about how to compile a non-cuda version of pytorch, which should have already been done in the OS specific sections right above this one.

The only difference between this section and the above ones is that it says to run `conda activate` first and then it has some details about OpenMP

Feels a bit redundant, and the CUDA section after this also needs a significant rewrite since it's written assuming a windows machine.

I don't understand the actual installation steps well enough to change the instructions just yet but in the long run I'd like to change the structure for the whole Install pytorch section to something like:

#### Install pytorch
There's two modes you can compile pytorch for, cpu-only computations and cuda-using computation mode.

If you want to use cuda, then first follow these instructions
To if you want to install a cpu-only version of pytorch, skip below to [this section](https://some.link)

File: README.md (hunk 6 above)
Comment (ZainRizvi) [reply README.md:158]: Note that the above section only applies to Linux environments

Comment (malfet) [reply README.md:158]: This sounds super out of date, conda have long since fixed this particular compiler issue. May be rephrase it to something like (latest patch revision of Python runtime is recommended)

**Review by malfet — APPROVED** [review]

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -4,14 +4,16 @@
...
+  - [Instructions](#instructions)
   - [Tips and Debugging](#tips-and-debugging)
 - [Nightly Checkout & Pull](#nightly-checkout--pull)
 - [Codebase structure](#codebase-structure)
 - [Unit testing](#unit-testing)
   - [Python Unit Testing](#python-unit-testing)
   - [Better local unit tests with `pytest`](#better-local-unit-tests-with-pytest)
   - [Local linting](#local-linting)
-  - [Running `mypy`](#running-mypy)
+    - [Running `mypy`](#running-mypy)
```
[the remark this answers is not in the document]
Comment (ZainRizvi) [reply CONTRIBUTING.md:16]: That's because I made it `Running mypy` a subsection of the `local linting` section.  MyPy is now run automatically by the linter, so the move seemed appropriate.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -125,8 +133,8 @@ python setup.py develop
 ```
 
 This mode will symlink the Python files from the current local source
-tree into the Python install.  Hence, if you modify a Python file, you
-do not need to reinstall PyTorch again and again.  This is especially
+tree into the Python install.  This way when you modify a Python file, you
+won't need to reinstall PyTorch again and again.  This is especially
```
[the remark this answers is not in the document]
Comment (ZainRizvi) [reply CONTRIBUTING.md:137]: I would argue that the more casual language is much easier for humans to comprehend and makes the document feel more familiar. (And most of the companies known for human-friendly language have shifted to this style as well)

It's unfortunate that we don't use it consistently yet, but this is a case in which I believe gradually transitioning to the better style is better than keeping everything inconsistently bad.  (If this was about coding style, I would agree with you about keeping each document consistent!)

File: README.md (hunk 4 above)
Comment (ZainRizvi) [reply README.md:182]: I like that idea, and it sounds similarly shaped to the ask in https://github.com/pytorch/pytorch/issues/72556

File: README.md (hunk 6 above)
Comment (ZainRizvi) [reply README.md:158]: Nice. Are you sure that conda has fixed the compiler issue that required python 3.7.6+?

If so, we can remove that error section and python version requirements completely

**Review by seemethere — APPROVED** [review]

File: README.md (hunk 4 above)
Comment (ragulpr) [reply README.md:182]: Attempting to give it a fix here! https://github.com/pytorch/pytorch/pull/91861
